### PR TITLE
p2p: add stream counter metric

### PR DESCRIPTION
**Pull request by gsora**
Adds a Prometheus metric to track the opened streams by protocol.

category: bug
ticket: #2277

### CODE REVIEW

File: p2p/ping.go [hunk 1]
```diff
@@ -79,50 +79,53 @@ func pingPeer(ctx context.Context, svc *ping.PingService, p peer.ID, callback fu
...
 ) {
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
 	// newPingChan creates a new stream and returns the ping result channel to listen for results
 	// and a close function to close the stream.
 	// Note: The only way to close a stream opened by ping service is to cancel the context.
-	newPingChan := func() (<-chan ping.Result, func()) {
-		pingCtx, pingCancel := context.WithCancel(ctx)
+	newPingChan := func(parentCtx context.Context) (<-chan ping.Result, func()) {
```
Comment (xenowits) [inline p2p/ping.go:89]: nit: can simply name `ctx`
```suggestion
	newPingChan := func(ctx context.Context) (<-chan ping.Result, func()) {
```

**Review by xenowits — APPROVED** [review]
LGTM!

File: p2p/ping.go [hunk 2]
```diff
@@ -79,50 +79,53 @@ func pingPeer(ctx context.Context, svc *ping.PingService, p peer.ID, callback fu
 	}
 }
 
-// pingPeerOnce starts a long lived ping connection with the peer and returns on first error.
+// pingPeerOnce starts a long-lived ping connection with the peer and returns on first error.
 func pingPeerOnce(ctx context.Context, svc *ping.PingService, p peer.ID,
```
Comment (corverroos) [inline p2p/ping.go:83]: all these closures and callbacks and stuff is very confusing, lets try and simplify into a simple loop. Take a step back, zoom out, think what the problem is, then solve it from scratch, we are building on top of previous stuff, we are digging the hole deeper and deeper

File: p2p/metrics.go [hunk 3]
```diff
@@ -56,6 +56,12 @@ var (
 		Help:      "Current number of libp2p connections by peer and type ('direct' or 'relay'). Note that peers may have multiple connections.",
 	}, []string{"peer", "type"})
 
+	peerStreamGauge = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "p2p",
+		Name:      "peer_streams",
+		Help:      "Current number of libp2p streams by peer and direction ('inbound' or 'outbound' or 'unknown').",
```
Comment (corverroos) [inline p2p/metrics.go:62]: help test needs to be updated

**Review by corverroos — CHANGES_REQUESTED** [review]

**Review by dB2510 — APPROVED** [review]

**Pull request merged**